src/ProducerEventProcessor/UploaderClientCaptureEventCollectorTest.cpp

```diff
@@ -0,0 +1,191 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include <gmock/gmock.h>
+#include <grpcpp/grpcpp.h>
+#include <grpcpp/support/channel_arguments.h>
+#include <gtest/gtest.h>
+
+#include <thread>
+
+#include "CaptureFile/CaptureFileOutputStream.h"
+#include "OrbitBase/Logging.h"
+#include "ProducerEventProcessor/UploaderClientCaptureEventCollector.h"
+#include "capture.pb.h"
+#include "services.grpc.pb.h"
+
+using orbit_grpc_protos::ClientCaptureEvent;
+
+namespace orbit_producer_event_processor {
+
+namespace {
+
+static constexpr const char* kAnswerString = "Test Interned String";
+static constexpr uint64_t kAnswerKey = 42;
+
+ClientCaptureEvent CreateInternedStringCaptureEvent(uint64_t key, const std::string& str) {
+  ClientCaptureEvent event;
+  event.mutable_interned_string()->set_key(key);
+  event.mutable_interned_string()->set_intern(str);
+  return event;
+}
+
+ClientCaptureEvent CreateCaptureFinishedEvent() {
+  ClientCaptureEvent event;
+  event.mutable_capture_finished()->set_status(orbit_grpc_protos::CaptureFinished::kSuccessful);
+  return event;
+}
+
+class UploaderClientCaptureEventCollectorTest : public testing::Test {
+ protected:
+  // Set `early_stop_event_count` to be >= 2 if we want to call `Stop()` early with producing only
+  // `early_stop_event_count` events. Here we require ">= 2" as an early stop will create at
+  // least two capture events: one `InternedStringCaptureEvent` created when stop_called_ becomes
+  // true, and one `CaptureFinishedEvent`.
+  void Start(size_t event_count, uint64_t produce_event_every_ms, uint64_t upload_events_every_ms,
+             size_t early_stop_event_count = 0) {
+    event_count_ = event_count;
+    produce_event_every_ms_ = produce_event_every_ms;
+    upload_events_every_ms_ = upload_events_every_ms;
+    early_stop_event_count_ = early_stop_event_count;
+
+    data_produce_thread_ =
+        std::thread(&UploaderClientCaptureEventCollectorTest::ProduceCaptureData, this);
+    data_upload_thread_ =
+        std::thread(&UploaderClientCaptureEventCollectorTest::UploadCaptureData, this);
+  }
+
+  void Stop() {
+    absl::MutexLock lock{&stop_called_mutex_};
+    stop_called_ = true;
+  }
+
+  void TearDown() override {
+    {
+      absl::MutexLock lock{&stop_called_mutex_};
+      CHECK(stop_called_);
```

Comment:
Don't use CHECK in a test fixture because this will crash the process and we won't get proper test results. You can use `ASSERT_TRUE` instead.